bundles/automation/org.eclipse.smarthome.automation.core/src/main/java/org/eclipse/smarthome/automation/core/internal/RuleRegistryImpl.java

```diff
@@ -461,87 +489,343 @@ public Boolean isEnabled(String ruleUID) {
+     * @return <code>true</code> if the type and value matching or <code>false</code> in the opposite.
+     */
+    private boolean checkType(Type type, Object configValue) {
+        switch (type) {
+            case TEXT:
+                return configValue instanceof String;
+            case BOOLEAN:
+                return configValue instanceof Boolean;
+            case INTEGER:
+                return configValue instanceof BigDecimal || configValue instanceof Integer
+                        || configValue instanceof Double && ((Double) configValue).intValue() == (Double) configValue;
+            case DECIMAL:
+                return configValue instanceof BigDecimal || configValue instanceof Double;
+        }
+        return false;
+    }
+
+    /**
+     * This method serves to replace module configuration references with the {@link Rule}s configuration values.
+     *
+     * @param modules the {@link Rule}'s modules, whose configuration values should be resolved.
+     * @param ruleConfiguration the {@link Rule}'s configuration values that should be resolve module configuration
+     *            values.
+     */
+    private void resolveModuleConfigReferences(List<? extends Module> modules, Map<String, ?> ruleConfiguration) {
+        if (modules != null) {
+            StringBuffer statusDescription = new StringBuffer();
+            for (Module module : modules) {
+                try {
+                    ReferenceResolver.updateConfiguration(module.getConfiguration(), ruleConfiguration, logger);
+                } catch (Exception e) {
+                    statusDescription.append(" in module[" + module.getId() + "]: " + e.getLocalizedMessage() + ";");
+                }
+            }
+            String statusDescriptionStr = statusDescription.toString();
+            if (!statusDescriptionStr.isEmpty()) {
+                throw new IllegalArgumentException(String.format("Incorrect configurations:%s", statusDescriptionStr));
+            }
+        }
     }
 
     @Override
     public void added(RuleTemplate element) {
         String templateUID = element.getUID();
         Set<String> rules = new HashSet<String>();
         synchronized (this) {
-            Set<String> rulesForResolving = mapTemplateToRules.remove(templateUID);
+            Set<String> rulesForResolving = mapTemplateToRules.get(templateUID);
             if (rulesForResolving != null) {
                 rules.addAll(rulesForResolving);
             }
         }
         for (String rUID : rules) {
-            Rule rule = get(rUID);
-            updateRuleByTemplate(getProvider(rule), rule);
-        }
-    }
-
-    private Provider<Rule> getProvider(Rule rule) {
-        for (Entry<Provider<Rule>, Collection<Rule>> entry : elementMap.entrySet()) {
-            if (entry.getValue().contains(rule)) {
-                return entry.getKey();
+            try {
+                update(resolveRuleByTemplate(get(rUID)));
+            } catch (Exception e) {
```

Comment:
We should not catch generic Exceptions. Can you make this more specific?

Reply:
OK